gammapy/estimators/energydependence.py

```diff
@@ -0,0 +1,259 @@
+
+        free_x, free_y = np.shape(
+            [result.parameters.free_parameters.names for result in results_src]
+        )
+        df_src = free_x * free_y
+
+        # Calculate the signal above the background
+        delta_ts_bkg_src = [
+            (n - a) for n, a in zip(results_bkg_total_stat, results_src_total_stat)
+        ]
+        df_bkg_src = df_src - df_bkg
+        sigma_ts_bkg_src = ts_to_sigma(delta_ts_bkg_src, df=df_bkg_src)
+
+        # Prepare results dictionary for signal above background
+        result_bkg_src = {}
+
+        result_bkg_src["Emin"] = self.energy_edges[:-1]
+        result_bkg_src["Emax"] = self.energy_edges[1:]
+        result_bkg_src["delta ts"] = [val for val in delta_ts_bkg_src]
+        result_bkg_src["df"] = [df_bkg_src] * self.num_energy_bands
+        result_bkg_src["significance"] = [elem for elem in sigma_ts_bkg_src]
+
+        return result_bkg_src
+
+    def estimate_energy_dependence(self, dataset):
+        """Estimate the potential of energy-dependent morphology.
+
+        Parameters
+        ----------
+        dataset : `~gammapy.datasets.MapDataset`
+            Input dataset to use.
+
+        Returns
+        -------
+        results : `dict`
+            Dict with results of the energy-dependence test.
+            * "delta_ts" : difference in ts between sliced fit and joint fit
+            * "df" : the degrees of freedom between sliced fit and joint fit
+            * "result" : the results for the sliced fit and joint fit
+        """
+        dataset.mask_fit = dataset.counts.geom.energy_mask(
+            energy_min=self.energy_edges[0], energy_max=None
+        )
+
+        # Calculate the individually sliced components
+        slices_src = Datasets()
+        for emin, emax in zip(self.energy_edges[:-1], self.energy_edges[1:]):
+            sliced_src = dataset.slice_by_energy(emin, emax)
+            bkg_sliced_model = FoVBackgroundModel(dataset_name=sliced_src.name)
+            sliced_src.models = [self.model.copy(), bkg_sliced_model]
+            slices_src.append(sliced_src)
+
+        results_src = []
+        for sliced in slices_src:
+            results_src.append(self.fit.run(sliced))
+
+        results_src_total_stat = [result.total_stat for result in results_src]
+        free_x, free_y = np.shape(
+            [result.parameters.free_parameters.names for result in results_src]
+        )
+        df_src = free_x * free_y
+
+        # Calculate the joint fit
+        parameters = self.model.spatial_model.parameters.free_parameters.names
+        slice0 = slices_src[0]
+        for slice_j in slices_src[1:]:
+            for param in parameters:
+                setattr(
+                    slice_j.models[0].spatial_model,
+                    param,
+                    slice0.models[0].spatial_model.parameters[param],
+                )
+        result_joint = self.fit.run(slices_src)
```

Comment:
Here you could do the fit on the original dataset, not the slices.
Only 3 fit are necessary:
-original dataset without the source
-original dataset with the source
(the two first can be done with TestStatisticNested  and read the stat value from the results)
-dataset per energy band with the source
and then you get TSdetection with or without energy-dependence, TSenergy-dependence.